pymagicc/input.py

```diff
@@ -0,0 +1,225 @@
+            'Could not find namelist within {}'.format(self.fname)
+        return nml_end, nml_start
+
+    def process_metadata(self, lines):
+        # TODO: replace with f90nml.reads when released
+        parser = f90nml.Parser()
+        nml = parser._readstream(lines, {})
+        metadata = {
+            k.split('_')[1]: nml['THISFILE_SPECIFICATIONS'][k]
+            for k in nml['THISFILE_SPECIFICATIONS']
+        }
+
+        return metadata
+
+    def process_data(self, stream, metadata):
+        """
+        Extract the tabulated data from a subset of the input file
+        :param stream: A Streamlike object (nominally StringIO) containing the
+            table to be extracted
+        :param metadata: Dictionary containing
+        :return: Tuple of a pd.DataFrame containing the data and a Dict
+            containing the units for each gas present in the output. The
+            pd.DataFrame columns are named using a MultiIndex of Gas and Region
+        """
+        raise NotImplementedError()
+
+
+class MAGICC6Reader(InputReader):
+    def process_data(self, stream, metadata):
+        gas = basename(self.fname).split('_')[1]
+        df = pd.read_csv(
+            stream,
+            skip_blank_lines=True,
+            delim_whitespace=True,
+            index_col=0,
+            engine="python")
+        # Convert to a columns to a MultiIndex
+        df.columns = [
+            [gas] * len(df.columns),
+            df.columns
+        ]
+        df.index.name = 'YEAR'
+
+        units = {
+            gas: metadata['units']
+        }
+        return df, units
+
+
+class MAGICC7Reader(InputReader):
+    def process_data(self, stream, metadata):
+        gases = stream.readline().split()
+        assert gases[0] == 'GAS'
+        gases = gases[1:]
+        stream.readline()  # Unused line
+        units = stream.readline().split()[1:]
+        regions = stream.readline().split()[1:]
+        index = pd.MultiIndex.from_arrays([gases, regions], names=['GAS', 'REGION'])
+        df = pd.read_csv(
+            stream,
+            skip_blank_lines=True,
+            delim_whitespace=True,
+            names=None,
+            header=None,
+            index_col=0)
+        df.index.name = 'YEAR'
+        df.columns = index
+
+        return df, self._extract_units(gases, units)
+
+    def _extract_units(self, gases, units):
+        combos = set(zip(gases, units))
+        result = {}
+        for v, u in combos:
+            if v not in result:
+                result[v] = u
+            else:
+                # this isn't expected to happen, but should check anyway
+                raise ValueError('Different units for {} in {}'.format(v, self.fname))
+
+        return result
+
+
+_file_types = {
+    'MAGICC6': MAGICC6Reader,
+    'MAGICC7': MAGICC7Reader,
+}
+
+
+def get_reader(fname):
+    with open(fname) as f:
+        lines = f.readlines()
+
+    # Infer the file type from the header
```

Comment:
we're going to have to be smarter about this but that's next step once we have a failing test